File: disciple/src/main/java/com/monster/disciple/transformer/AbsErrorCheckTransformer.java
package com.monster.disciple.transformer;

import android.content.Context;
import android.text.TextUtils;

import com.monster.disciple.CacheBuilder;
import com.monster.disciple.entity.CacheEntity;
import com.monster.disciple.response.BaseResponse;
import com.monster.disciple.util.Util;

import java.lang.ref.WeakReference;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Response;

public abstract class AbsErrorCheckTransformer<T extends Response<R>, R extends BaseResponse>
        implements ObservableTransformer<T, R>, IErrorCheckTransformer {

    protected WeakReference<Context> mContextRef;
    private CacheBuilder mCacheBuilder;

    public AbsErrorCheckTransformer(final Context context) {
        mContextRef = new WeakReference<>(context);
    }

    public AbsErrorCheckTransformer(final Context context, final CacheBuilder cacheBuilder) {
        mContextRef = new WeakReference<>(context);
        mCacheBuilder = cacheBuilder;
    }

    @Override
    public ObservableSource<R> apply(Observable<T> upstream) {
        if (mCacheBuilder == null || TextUtils.isEmpty(mCacheBuilder.getCacheKey())) {
            return upstream
                    .map(Util.<T, R>newInstanceFunction(this))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
        } else {
            return upstream
                    .map(Util.<T, R>newInstanceFunction(this))
                    .map(Util.<R>saveCacheFunction(mContextRef.get(), mCacheBuilder.getCacheKey()))
                    .startWith(
                            Observable
                                    .just(mCacheBuilder.getCacheKey())
                                    .map(Util.<R>getCacheFunction(mContextRef.get(), mCacheBuilder.getType()))
                                    .filter(new Predicate<CacheEntity<R>>() {
                                        @Override
                                        public boolean test(CacheEntity<R> rCacheEntity) {
                                            return !rCacheEntity.isNil();
                                        }
                                    })
                                    .map(new Function<CacheEntity<R>, R>() {
                                        @Override
                                        public R apply(CacheEntity<R> rCacheEntity) {
                                            return rCacheEntity.getResponse();
                                        }
                                    }))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread());
        }
    }
}

File: disciple/src/main/java/com/monster/disciple/transformer/AbsErrorCheckTransformer.kt
package com.monster.disciple.transformer

import android.content.Context
import android.text.TextUtils
import com.monster.disciple.CacheBuilder
import com.monster.disciple.response.BaseResponse
import com.monster.disciple.util.getCacheFunction
import com.monster.disciple.util.newInstanceFunction
import com.monster.disciple.util.saveCacheFunction
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import retrofit2.Response
import java.lang.ref.WeakReference

abstract class AbsErrorCheckTransformer<T : Response<R>, R : BaseResponse<*>>(context: Context?) : ObservableTransformer<T, R>, IErrorCheckTransformer {

    private var mContextRef: WeakReference<Context>? = WeakReference<Context>(context)

    private var mCacheBuilder: CacheBuilder? = null

    constructor(context: Context?, cacheBuilder: CacheBuilder?) : this(context) {
        mCacheBuilder = cacheBuilder
    }

    override fun apply(upstream: Observable<T>): ObservableSource<R> {
        return if (mCacheBuilder == null || TextUtils.isEmpty(mCacheBuilder?.cacheKey)) {
            upstream
                    .map(newInstanceFunction<T,R>(this))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        } else {
            upstream
                    .map(newInstanceFunction<T,R>(this))
                    .map(saveCacheFunction<R>(mContextRef?.get(), mCacheBuilder?.cacheKey))
                    .startWith(
                            Observable
                                    .just<String>(mCacheBuilder?.cacheKey)
                                    .map(getCacheFunction<R>(mContextRef?.get(), mCacheBuilder?.type))
                                    .filter { rCacheEntity -> !rCacheEntity.isNil() }
                                    .map { rCacheEntity -> rCacheEntity.response })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }
}

